cpp/src/groupby/new_groupby.cu

```diff
@@ -190,3 +192,230 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
+   *    GDF_PANDAS_STYLE = Nulls are ignored in group by keys (Pandas style),
+   *    GDF_SQL_STYLE = Nulls are treated as values in group by keys where NULL == NULL (SQL style)>
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_by_without_aggregations(gdf_size_type num_data_cols,
+                                            gdf_column** data_cols_in,
+                                            gdf_size_type num_groupby_cols,
+                                            gdf_index_type * groupby_col_indices,
+                                            gdf_column** data_cols_out,
+                                            gdf_column* group_start_indices,
+                                            gdf_context* ctxt)      
+{
+  GDF_REQUIRE((nullptr != data_cols_in), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != data_cols_in[0]), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != data_cols_out), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != data_cols_out[0]), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((num_data_cols > 0), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((num_groupby_cols > 0), GDF_DATASET_EMPTY);
+  GDF_REQUIRE((nullptr != groupby_col_indices), GDF_DATASET_EMPTY);
+
+
+  int32_t nrows = data_cols_in[0]->size;
+
+  // setup for order by call
+  bool group_by_keys_contain_nulls = false;
+  std::vector<gdf_column*> orderby_cols_vect(num_groupby_cols);
+  for (int i = 0; i < num_groupby_cols; i++){
+    orderby_cols_vect[i] = data_cols_in[groupby_col_indices[i]];
+    group_by_keys_contain_nulls = group_by_keys_contain_nulls || orderby_cols_vect[i]->null_count > 0;
+  }
+
+  rmm::device_vector<int32_t> sorted_indices(nrows);
+  gdf_column sorted_indices_col;
+  gdf_error status = gdf_column_view(&sorted_indices_col, (void*)(sorted_indices.data().get()), 
+                            nullptr, nrows, GDF_INT32);
+  if (status != GDF_SUCCESS)
+    return status;
+
+  if (ctxt->flag_groupby_include_nulls == GDF_SQL_STYLE || !group_by_keys_contain_nulls){  // SQL style
+  // run order by and get new sort indexes
+    status = gdf_order_by(&orderby_cols_vect[0],             //input columns
+                          nullptr,
+                          num_groupby_cols,                //number of columns in the first parameter (e.g. number of columsn to sort by)
+                          &sorted_indices_col,            //a gdf_column that is pre allocated for storing sorted indices
+                          ctxt);
+    if (status != GDF_SUCCESS)
+      return status;
+
+    // run gather operation to establish new order
+    std::unique_ptr< gdf_table<int32_t> > table_in{new gdf_table<int32_t>{num_data_cols, data_cols_in}};
+    std::unique_ptr< gdf_table<int32_t> > table_out{new gdf_table<int32_t>{num_data_cols, data_cols_out}};
+
+    status = table_in->gather<int32_t>(sorted_indices, *table_out.get());
```

Comment:
You'll want to use the new `gather` that is a free function instead of a member of `gdf_table`.

The PR is almost wrapped up and should be merged soon.
https://github.com/rapidsai/cudf/pull/577